crates/miden-lib/asm/note_scripts/P2IDE.masm

```diff
@@ -0,0 +1,201 @@
+use.miden::account
+use.miden::note
+use.miden::tx
+use.miden::contracts::wallets::basic->wallet
+
+# ERRORS
+# =================================================================================================
+
+const.ERR_P2IDE_WRONG_NUMBER_OF_INPUTS="P2IDE note expects exactly 4 note inputs"
+
+const.ERR_P2IDE_RECLAIM_ACCT_IS_NOT_SENDER="failed to reclaim P2IDE note because the reclaiming account is not the sender"
+
+const.ERR_P2IDE_RECLAIM_HEIGHT_NOT_REACHED="failed to reclaim P2IDE note because the reclaim block height is not reached yet"
+
+const.ERR_P2IDE_RECLAIM_DISABLED="P2IDE reclaim is disabled"
+
+const.ERR_P2IDE_TIMELOCK_HEIGHT_NOT_REACHED="failed to consume P2IDE note because the note is still timelocked"
+
+# CONSTANTS
+# =================================================================================================
+
+const.CURRENT_BLOCK_HEIGHT_MEM_ADDR=5
+
+# HELPER PROCEDURES
+# =================================================================================================
+
+#! Helper procedure to add all assets of a note to an account.
+#!
+#! Inputs:  []
+#! Outputs: []
+proc.add_note_assets_to_account
+    push.0 exec.note::get_assets
+    # => [num_of_assets, 0 = ptr, ...]
+
+    # compute the pointer at which we should stop iterating
+    mul.4 dup.1 add
+    # => [end_ptr, ptr, ...]
+
+    # pad the stack and move the pointer to the top
+    padw movup.5
+    # => [ptr, 0, 0, 0, 0, end_ptr, ...]
+
+    # compute the loop latch
+    dup dup.6 neq
+    # => [latch, ptr, 0, 0, 0, 0, end_ptr, ...]
+
+    while.true
+        # => [ptr, 0, 0, 0, 0, end_ptr, ...]
+
+        # save the pointer so that we can use it later
+        dup movdn.5
+        # => [ptr, 0, 0, 0, 0, ptr, end_ptr, ...]
+
+        # load the asset
+        mem_loadw
+        # => [ASSET, ptr, end_ptr, ...]
+
+        # pad the stack before call
+        padw swapw padw padw swapdw
+        # => [ASSET, pad(12), ptr, end_ptr, ...]
+
+        # add asset to the account
+        call.wallet::receive_asset
+        # => [pad(16), ptr, end_ptr, ...]
+
+        # clean the stack after call
+        dropw dropw dropw
+        # => [0, 0, 0, 0, ptr, end_ptr, ...]
+
+        # increment the pointer and compare it to the end_ptr
+        movup.4 add.4 dup dup.6 neq
+        # => [latch, ptr+4, ASSET, end_ptr, ...]
+    end
+
+    # clear the stack
+    drop dropw drop
+end
+
+#! Helper procedure to check if the P2IDE note is unlocked.
+#!
+#! Inputs:  [timelock_block_height, reclaim_block_height, target_account_id_prefix, target_account_id_suffix]
+#! Outputs: [reclaim_block_height, target_account_id_prefix, target_account_id_suffix]
+proc.verify_unlocked
+    # get current block number and store in memory for later
+    exec.tx::get_block_number dup mem_store.CURRENT_BLOCK_HEIGHT_MEM_ADDR
```

Comment:
Another option (instead of storing current block height to memory) would be to return it from the procedure. Would improve readability a bit.